### PR TITLE
Add `forwarders` to `dns::zone`; deprecate `allow_forwarder`

### PR DESCRIPTION
**Pull request by ghost**
This change adds the `forwarders` parameter to `dns::zone`, deprecating
`allow_forwarder` in favor of `forwarders`.  `forwarders` extends
`allow_forwarder` by allowing both an *empty* forwarders list as well as
an *undefined* forwarders list (the former meaning that the zone does not
forward, while the latter leaves the global forwarders option in place):

> [*forwarders*]
> An array of IP addresses and optional port numbers to which queries
> for this zone will be forwarded (based on the *forward_policy*
> setting).  If the optional port number is included, it must be
> separated from the IP address by the word `port` - for example,
> `[ '192.168.100.102 port 1234' ]`.  If passed an empty array or the
> boolean value `false`, the zone will not forward.  If passed `true`
> or left undefined, the zone will use the global forwarders defined
> in `dns::server::options`.
> *Note* - this parameter deprecates and should be used in place of
> the *allow_forwarder* parameter.  If both parameters are passed in,
> only *forwarders* will take effect.

This addresses Issue #204

### CODE REVIEW

**Comment by jearls** [comment]
Note - as this PR deprecates a parameter, SemVer requires a minor version bump if/when this is published.

File: manifests/zone.pp [hunk 1]
```diff
@@ -175,12 +189,46 @@
...
+      if $forwarders {
+        $zone_forwarders = undef
+      } else {
+        $zone_forwarders = []
+      }
+    } else {
+      validate_array($forwarders)
+      $zone_forwarders = $forwarders
+    }
+  } else {
```
Comment (apeschel) [inline manifests/zone.pp]: Not sure about the style on this module, but this would read cleaner for me if lines 224-225 were replaced with `elsif`.

**Review by apeschel — CHANGES_REQUESTED** [review]
Looks good, just one style suggestion.

**Comment by jjthiessen** [comment]
I have a similar change in https://github.com/jjthiessen/puppet-dns/commit/1e2d3357b089b0610c654301cd4b943ae56be928, except I went with [option 2](https://github.com/ajjahn/puppet-dns/issues/204#issuecomment-284132399) and extended forwarders to _slave_ and _stub_ zones (see the [zone statement grammar](https://ftp.isc.org/isc/bind9/cur/9.9/doc/arm/Bv9ARM.ch06.html#zone_statement_grammar) for justification).

If this (_option 3_) is the preferred approach, can we augment _templates/zone.erb_ to support _slave_ and _stub_ zones and get this merged?

If it's preferable to keep the empty forwarder feature addition, and _slave_ and _stub_ forwarder support in separate PRs, I can submit a follow-up PR (it likely makes sense to merge this first as (either way) one will have to be rebased on the other).

Is there any particular hold-up on this PR?

**Comment by jjthiessen** [comment]
As an aside, https://github.com/jjthiessen/puppet-dns/commits/add-forwarder-support-for-slave-and-stub-zones is this branch with the practically trivial addition of slave and stub forwarder support.

**Pull request closed without merging**